{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3a6cd46c",
   "metadata": {},
   "source": [
    "\n",
    "# Bioinformatic analysis of the role of GemC1 in the regulation of transcription and chromatin dynamics by BAF complex during differentiation of multiciliated ependymal cells\n",
    "\n",
    "**Project:** Bioinformatic analysis of the role of GemC1 in chromatin dynamics.  \n",
    "**Assay:** CUT&RUN (Brg1)  \n",
    "**Organism:** *Mus musculus* (mm10)\n",
    "\n",
    "This notebook documents the full CUT&RUN bioinformatics pipeline used in my master's thesis project titled “Bioinformatic analysis of the role of GemC1 in the regulation of transcription and chromatin dynamics by BAF complex during differentiation of multiciliated ependymal cells”.\n",
    "\n",
    "The full text can be found in the following link: https://nemertes.library.upatras.gr/items/1781a0e5-00d6-480e-abc1-7257d588e0ce"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e084e0af",
   "metadata": {},
   "source": [
    "\n",
    "## 1. Software and Environment\n",
    "\n",
    "### Linux tools\n",
    "- Bowtie2 (>= 2.3.4.3)\n",
    "- samtools (>= 1.10)\n",
    "- bedtools (>= 2.29.1)\n",
    "- SEACR (>= 1.3)\n",
    "- deepTools\n",
    "- MACS2\n",
    "\n",
    "### R packages\n",
    "- dplyr, stringr\n",
    "- ggplot2, viridis\n",
    "- GenomicRanges\n",
    "- DESeq2\n",
    "- ggpubr, corrplot\n",
    "- chromVAR\n",
    "- data.table\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4346c791",
   "metadata": {},
   "source": [
    "\n",
    "## 2. Project Structure and Sample Metadata\n",
    "\n",
    "Samples follow the naming convention:\n",
    "\n",
    "`sampleID_antibody_genotype_replicate`\n",
    "\n",
    "Files Used:\n",
    "- `80.7_Brg1_KO_rep1.bam`\n",
    "- `80.10_Brg1_WT_rep1.bam`\n",
    "- `80.11_Brg1_KO_rep2.bam`\n",
    "- `80.13_Brg1_WT_rep2.bam`\n",
    "- `80.1_IgG_KO_rep1.bam`\n",
    "- `80.18_IgG_WT_rep1.bam`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6796411e",
   "metadata": {},
   "source": [
    "\n",
    "## 3. Alignment and Quality Control\n",
    "\n",
    "BAM files were inspected, converted, and aligned.\n",
    "The following command gives us further information on the bam files:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81ac6487",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "samtools view $projPath/alignment/bam/${sampleID}_${antibody}_${genotype}_${replicate}.bam | less -S"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d98b308d",
   "metadata": {},
   "source": [
    "Investigation on the previous step showed that 80.11 bam file was truncated, so we converted the corresponding cram file into bam format:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cf00335",
   "metadata": {},
   "outputs": [],
   "source": [
    "samtools view -b -T $projPath/alignment/mm10/mm10.fa -o \n",
    "$projPath/alignment/bam/${sampleID}_${antibody}_${genotype}_${replicate}.bam  \n",
    "$projPath/alignment/cram/${sampleID}_${antibody}_${genotype}_${replicate}.cram"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae983a81",
   "metadata": {},
   "source": [
    "Convert bam into fastq file format:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3021301",
   "metadata": {},
   "outputs": [],
   "source": [
    "bedtools bamtofastq -i $projPath/alignment/bam/${sampleID}_${antibody}_${genotype}_${replicate}.bam \\\n",
    "                      -fq $projPath/data/fastq_files/${sampleID}_${antibody}_${genotype}_${replicate}_R1.fq \\\n",
    "                      -fq2 $projPath/data/fastq_files/${sampleID}_${antibody}_${genotype}_${replicate}_R2.fq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e048f715",
   "metadata": {},
   "source": [
    "Run bowtie2 to perform the alignment to the mus musculus genome mm10 and get the summary .txt file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67fa74f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "cores=8\n",
    "ref=\"/path/to/mm10/index/\"\n",
    "bowtie2 --end-to-end --very-sensitive --no-mixed --no-discordant --phred33 -I 10 -X 700 -p ${cores} -x ${ref}/mm10 -1 ${projPath}/data/fastq_files/${sampleID}_${antibody}_${genotype}_${replicate}_R1.fq -2 ${projPath}/data/fastq_files/${sampleID}_${antibody}_${genotype}_${replicate}_R2.fq -S ${projPath}/alignment/sam/${sampleID}_${antibody}_${genotype}_${replicate}_bowtie2.sam &> ${projPath}/alignment/sam/bowtie2_summary/${sampleID}_${antibody}_${genotype}_${replicate}_bowtie2.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5752ac7a",
   "metadata": {},
   "source": [
    "Convert bam into bed file format:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daf68ae0",
   "metadata": {},
   "outputs": [],
   "source": [
    "bedtools bamtobed -bedpe -i \n",
    "$projPath/alignment/bam/${sampleID}_${antibody}_${genotype}_${replicate}.bam >$projPath/alignment/bed/${sampleID}_${antibody}_${genotype}_${replicate}.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f974f90d",
   "metadata": {},
   "source": [
    "Delete \"chr. -1 -1\" from 80.11:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e266a555",
   "metadata": {},
   "outputs": [],
   "source": [
    "grep -v \"\\.\" $projPath/alignment/bed/${sampleID}_${antibody}_${genotype}_${replicate}.bed > $projPath/alignment/bed/${sampleID}_${antibody}_${genotype}_${replicate}.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d194425c",
   "metadata": {},
   "source": [
    "Keep the read pairs that are on the same chromosome and fragment length less than 1000bp:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b02328d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "awk '$1==$4 && $6-$2 < 1000 {print $0}' $projPath/alignment/bed/${sampleID}_${antibody}_${genotype}_${replicate}.bed >$projPath/alignment/bed/${sampleID}_${antibody}_${genotype}_${replicate}.clean.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17e1627e",
   "metadata": {},
   "source": [
    "Only extract the fragment related columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df0e6a92",
   "metadata": {},
   "outputs": [],
   "source": [
    "cut -f 1,2,6 $projPath/alignment/bed/${sampleID}_${antibody}_${genotype}_${replicate}.clean.bed | sort -k1,1 -k2,2n -k3,3n  >$projPath/alignment/bed/${sampleID}_${antibody}_${genotype}_${replicate}.fragments.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6549e0fc",
   "metadata": {},
   "source": [
    "Visualize the sequencing depth and alignment results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8594a43",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "alignResult = c()\n",
    "for(sample in sampleList2){\n",
    "  alignRes = read.table(paste0(projPath, \"/alignment/sam/bowtie2_summary/\", sample, \"_bowtie2.txt\"), header = FALSE, fill = TRUE)\n",
    "  alignRate = substr(alignRes$V1[6], 1, nchar(as.character(alignRes$V1[6]))-1)\n",
    "  sampleInfo = strsplit(sample, \"_\")[[1]]\n",
    "  alignResult = data.frame(Sample = paste0(sampleInfo[1], \"_\", sampleInfo[2], \"_\", sampleInfo[3], \"_\", sampleInfo[4]), \n",
    "                           Replicate = sampleInfo[4],\n",
    "                           Genotype = sampleInfo[3],\n",
    "                           SequencingDepth = alignRes$V1[1] %>% as.character %>% as.numeric, \n",
    "                           MappedFragNum_mm10 = alignRes$V1[4] %>% as.character %>% as.numeric + alignRes$V1[5] %>% as.character %>% as.numeric, \n",
    "                           AlignmentRate_mm10 = alignRate %>% as.numeric)  %>% rbind(alignResult, .)\n",
    "}\n",
    "alignResult %>% mutate(AlignmentRate_mm10 = paste0(AlignmentRate_mm10, \"%\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5cb8794-4d2b-407f-ac98-87c5497d12ee",
   "metadata": {},
   "source": [
    "![](images/1.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5424db55",
   "metadata": {},
   "source": [
    "It seems that 80.11 has a higher sequencing depth and a lower alignment rate indicating that sample quality was compromised, probably due to technical issues, during sample preparation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a3b21e1",
   "metadata": {},
   "source": [
    "\n",
    "## 4. Sequencing Depth and Alignment Metrics\n",
    "\n",
    "Bowtie2 alignment summaries were parsed to extract:\n",
    "- Sequencing depth\n",
    "- Number of mapped fragments\n",
    "- Alignment rate\n",
    "\n",
    "ggplot package was used to generate graphs to visualize the sequencing depth and alignment results:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7afa1c83",
   "metadata": {},
   "outputs": [],
   "source": [
    "figA = alignResult %>% ggplot(aes(x = Sample, y = SequencingDepth/1000000, fill = Sample)) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = Genotype), position = position_jitter(0.15)) +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1, end = 0.9, option = \"magma\", alpha = 0.8) +\n",
    "  scale_color_viridis(discrete = TRUE, begin = 0.1, end = 0.9) +\n",
    "  theme_bw(base_size = 13) +\n",
    "  ylab(\"Sequencing Depth per Million\") +\n",
    "  xlab(\"\") + \n",
    "  ggtitle(\"A. Sequencing Depth\")\n",
    "\n",
    "figB = alignResult %>% ggplot(aes(x = Sample, y = MappedFragNum_mm10/1000000, fill = Sample)) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = Genotype), position = position_jitter(0.15)) +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1, end = 0.9, option = \"magma\", alpha = 0.8) +\n",
    "  scale_color_viridis(discrete = TRUE, begin = 0.1, end = 0.9) +\n",
    "  theme_bw(base_size = 13) +\n",
    "  ylab(\"Mapped Fragments per Million\") +\n",
    "  xlab(\"\") +\n",
    "  ggtitle(\"B. Alignable Fragment (mm10)\")\n",
    "\n",
    "figC = alignResult %>% ggplot(aes(x = Sample, y = AlignmentRate_mm10, fill = Sample)) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = Genotype), position = position_jitter(0.15)) +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1, end = 0.9, option = \"magma\", alpha = 0.8) +\n",
    "  scale_color_viridis(discrete = TRUE, begin = 0.1, end = 0.9) +\n",
    "  theme_bw(base_size = 13) +\n",
    "  ylab(\"% of Mapped Fragments\") +\n",
    "  xlab(\"\") +\n",
    "  ggtitle(\"C. Alignment Rate (mm10)\")\n",
    "\n",
    "ggarrange(figA, figB, figC, ncol = 2, nrow=2, common.legend = TRUE, legend=\"bottom\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72a7d829-6597-4b16-a67c-0fafb3592974",
   "metadata": {},
   "source": [
    "![Quality analysis of alignment on mm10. Plots for A. Sequencing Depth, B. Alignable Rate (mm10) and C. Alignment Rate (mm10)](images/2.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdef5ad5",
   "metadata": {},
   "source": [
    "Sequencing depth of sample 80.11 is high but the percentage of DNA corresponding to the mm10 genome is significantly lower than the other samples.\n",
    "This leads us to the conclusion that probably this sample contains a smaller amount of DNA which was read more times. \n",
    "\n",
    "This is probably leading to many copies of the adapter dimers which were amplified multiple times but cannot be sequenced in the genome. This is a typical case of a low complexity library.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4455a60",
   "metadata": {},
   "source": [
    "\n",
    "## 5. Replicate Reproducibility\n",
    "\n",
    "Genome-wide reproducibility was assessed by:\n",
    "- Binning the genome into 500 bp windows\n",
    "- Counting fragments per bin\n",
    "- Calculating Pearson correlations between samples\n",
    "\n",
    "It seems there is not very good reproducibility between the replicates for each condition that could be due to technical reasons or biological heterogeneity between the primary cultures of RGCs used."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32da161f",
   "metadata": {},
   "source": [
    "\n",
    "## 6. Peak Calling\n",
    "\n",
    "In order to identify the regions of the genome that are significantly overrepresented in our mapped files and constitute the peaks two peak callers were tested:\n",
    "- *SEACR*\n",
    "- *MACS2*\n",
    "\n",
    "First, add \"chr\" prefix on the first column in bed files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4a1eb16",
   "metadata": {},
   "outputs": [],
   "source": [
    "awk '{print \"chr\" $0}' OFS=\\t  $projPath/alignment/bed/${sampleID}_${antibody}_${genotype}_${replicate}.fragments.bed >$projPath/alignment/bed/${sampleID}_${antibody}_${genotype}_${replicate}_prefix.fragments.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e5281e4",
   "metadata": {},
   "source": [
    "Create bedgraph files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5e3c46b",
   "metadata": {},
   "outputs": [],
   "source": [
    "chromSize=\"/path/to/mm10.chrom.size\"\n",
    "bedtools genomecov -bg -i $projPath/alignment/bed/${sampleID}_${antibody}_${genotype}_${replicate}_prefix.fragments.bed -g $chromSize > $projPath/alignment/bedgraph/${sampleID}_${antibody}_${genotype}_${replicate}.fragments.bedgraph"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d83b1933",
   "metadata": {},
   "source": [
    "Perform peak calling with SEACR:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73f43197",
   "metadata": {},
   "outputs": [],
   "source": [
    "seacr=' path/to/SEACR_1.3.sh'\n",
    "mkdir -p $projPath/peakCalling/SEACR\n",
    "\n",
    "bash $seacr $projPath/alignment/bedgraph/${sampleID}_${antibody}_${genotype}_${replicate}.fragments.bedgraph \\\n",
    "     $projPath/alignment/bedgraph/<INSERT CONTROL HERE>.fragments.bedgraph \\\n",
    "     non stringent $projPath/peakCalling/SEACR/${sampleID}_${antibody}_${genotype}_${replicate}_seacr_control.peaks\n",
    "\n",
    "bash $seacr $projPath/alignment/bedgraph/${sampleID}_${antibody}_${genotype}_${replicate}.fragments.bedgraph 0.01 non stringent $projPath/peakCalling/SEACR/${sampleID}_${antibody}_${genotype}_${replicate}_seacr_top0.01.peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "715ea279",
   "metadata": {},
   "source": [
    "Calculate number of top 0.01 peaks called:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae592e83",
   "metadata": {},
   "outputs": [],
   "source": [
    "peakN = c()\n",
    "peakWidth = c()\n",
    "for(sample in sampleList2){\n",
    "  sampleInfo = strsplit(sample, \"_\")[[1]]\n",
    "      peakInfo = read.table(paste0(projPath, \"/peakCalling/SEACR/\", sample, \"_seacr_top0.01.peaks.stringent.bed\"), header = FALSE, fill = TRUE)  %>% mutate(width = abs(V3-V2))\n",
    "      peakN = data.frame(peakN = nrow(peakInfo), Sample = paste0(sampleInfo[1], \"_\", sampleInfo[2], \"_\", sampleInfo[3], \"_\", sampleInfo[4]), Genotype = sampleInfo[3], Replicate = sampleInfo[4]) %>% rbind(peakN, .)\n",
    "      peakWidth = data.frame(width = peakInfo$width, Sample = paste0(sampleInfo[1], \"_\", sampleInfo[2], \"_\", sampleInfo[3], \"_\", sampleInfo[4]), Genotype = sampleInfo[3], Replicate = sampleInfo[4])  %>% rbind(peakWidth, .)\n",
    "}\n",
    "peakN %>% select(Sample, Genotype, Replicate, peakN)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "507d5dba-07d3-4fbe-8880-af5aa69910ce",
   "metadata": {},
   "source": [
    "![](images/3.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e5c4941",
   "metadata": {},
   "source": [
    "Remove \"chr\" prefix on the first column in bed files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db8751bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "awk '{sub(/^chr/, \"\", $1); print}' $projPath/peakCalling/SEACR/new_notation/WT_rep2_seacr_top0.01.peaks.stringent.bed > $projPath/peakCalling/SEACR/new_notation/no_prefix/WT_rep2_seacr_top0.01.peaks.stringent_no_prefix.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0ae700c",
   "metadata": {},
   "source": [
    "Calculate FRagment proportion in Peaks regions (FRiPs):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36c39c49",
   "metadata": {},
   "outputs": [],
   "source": [
    "inPeakData = c()\n",
    "  for(rep in repL){\n",
    "    for(type in genotypes){\n",
    "      peakRes = read.table(paste0(projPath, \"/peakCalling/SEACR/new_notation/no_prefix/\", type, \"_\", rep, \"_seacr_top0.01.peaks.stringent_no_prefix.bed\"), header = FALSE, fill = TRUE)\n",
    "peak.gr = GRanges(seqnames = peakRes$V1, IRanges(start = peakRes$V2, end = peakRes$V3), strand =\"*\")\n",
    "      bamFile = paste0(bamDir2, type, \"_\", rep, \".bam\")\n",
    "      fragment_counts <- getCounts(bamFile, peak.gr, paired = TRUE, by_rg = FALSE, format = \"bam\")\n",
    "      inPeakN = counts(fragment_counts)[,1] %>% sum\n",
    "      inPeakData = rbind(inPeakData, data.frame(inPeakN = inPeakN, Replicate = rep, Genotype = type))\n",
    "  }\n",
    "}\n",
    "frip = left_join(inPeakData, alignResult, by = c(\"Genotype\", \"Replicate\")) %>% mutate(frip = inPeakN/MappedFragNum_mm10 * 100)\n",
    "frip %>% select(Genotype, Replicate, SequencingDepth, MappedFragNum_mm10, AlignmentRate_mm10, FragInPeakNum = inPeakN, FRiPs = frip)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7f12dc8-d368-43f4-92ee-91b257722f11",
   "metadata": {},
   "source": [
    "![](images/4.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9f8f326",
   "metadata": {},
   "source": [
    "Visualization of peaks quality analysis results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dba217c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig2A = peakN %>% ggplot(aes(x = Sample, y = peakN, fill = Sample)) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = Genotype), position = position_jitter(0.15)) +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1, end = 0.55, option = \"magma\", alpha = 0.8) +\n",
    "  scale_color_viridis(discrete = TRUE, begin = 0.1, end = 0.9) +\n",
    "  theme_bw(base_size = 13) +\n",
    "  ylab(\"Number of Peaks\") +\n",
    "  xlab(\"\")\n",
    "\n",
    "fig2B = peakWidth %>% ggplot(aes(x = \"\", y = width, fill = Genotype)) +\n",
    "  geom_violin() +\n",
    "  facet_grid(Genotype~Sample) +\n",
    "  facet_grid(~factor(Sample, levels=c(\"80.10_Brg1_WT_rep1\", \"80.13_Brg1_WT_rep2\", \"80.7_Brg1_KO_rep1\", \"80.11_Brg1_KO_rep2\")))+\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1, end = 0.55, option = \"magma\", alpha = 0.8) +\n",
    "  scale_color_viridis(discrete = TRUE, begin = 0.1, end = 0.9) +\n",
    "  scale_y_continuous(trans = \"log\", breaks = c(400, 3000, 22000)) +\n",
    "  theme_bw(base_size = 13) +\n",
    "  ylab(\"Width of Peaks\") +\n",
    "  xlab(\"\")\n",
    "\n",
    "fig2C = frip %>% ggplot(aes(x = Sample, y = frip, fill = Sample, label = round(frip, 2))) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = Genotype), position = position_jitter(0.15)) +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1, end = 0.55, option = \"magma\", alpha = 0.8) +\n",
    "  scale_color_viridis(discrete = TRUE, begin = 0.1, end = 0.9) +\n",
    "  theme_bw(base_size = 13) +\n",
    "  ylab(\"% of Fragments in Peaks\") +\n",
    "  xlab(\"\")\n",
    "ggarrange(fig2A, fig2B, fig2C, ncol = 2, nrow=2, common.legend = TRUE, legend=\"bottom\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1f3091a",
   "metadata": {},
   "source": [
    "![](images/5.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5c66218",
   "metadata": {},
   "source": [
    "Peak width is similar between the samples and sample 80.11 has a significantly smaller number of peaks.\n",
    "If we look at the percentage of DNA assigned to peaks, we can see that the normal samples have a similar percentage while 80.11 has a significantly higher percentage. So, while the 80.11 sample had a lower percentage of DNA mapped to the genome, this DNA contained many peaks.\n",
    "Overall, we conclude that probably this sample had a lower yield due to technical problems during the isolation and processing of DNA from this sample.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abefec53",
   "metadata": {},
   "source": [
    "\n",
    "## 7. Visualization\n",
    "\n",
    "Create bigwig files:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae782959",
   "metadata": {},
   "outputs": [],
   "source": [
    "bamCoverage -b $projPath/alignment/bam/${sampleID}_${antibody}_${genotype}_${replicate}.bam -o $projPath/alignment/bigwig/${sampleID}_${antibody}_${genotype}_${replicate}_raw.bw"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10e44254",
   "metadata": {},
   "source": [
    "Visual inspection of SEACR-called peaks in combination with genome coverage bw files reveals that in most cases, SEACR has called multiple false positive peaks. This is probably because the Brg1 genomic binding pattern is a combination of sharp and broad peaks that cannot be fully captured by the SEACR algorithm. For this reason, we also called peaks using the MACS2 package.\n",
    "\n",
    "To perform peak calling with MACS2 we uploaded the bam files on Galaxy platform which is a scientific workflow, data integration and data and analysis persistence and publishing platform. We used the default options for paired end bam files. The outputs of MACS2 were bed files that contain all peaks called from the corresponding bam files.\n",
    "\n",
    "Visual inspection of MACS2 called peaks in combination with genome coverage files reveals a tight agreement between the called peaks and the binding of Brg1, further suggesting that MACS2 has greater power in calling peaks from a binding pattern of both sharp and broad peaks, such as the pattern of Brg1 binding.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c34ac43",
   "metadata": {},
   "source": [
    "The representative loci of Brg1 sites lost/reduced in GemC1 KO cells can be seen below:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97f093d7",
   "metadata": {},
   "source": [
    "![](images/6.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29db9179-8a68-422b-8942-96d453b9fefb",
   "metadata": {},
   "source": [
    "![](images/7.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c63581f0-2e90-45dc-b684-7dd927829f7f",
   "metadata": {},
   "source": [
    "Next, the representative loci of Brg1 sites gained in GemC1 KO cells:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ab5e9a0-601d-4ee8-bed7-f051b8959631",
   "metadata": {},
   "source": [
    "![](images/8.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79404260-3fa2-493b-9b2a-0ed70bfb6fbf",
   "metadata": {},
   "source": [
    "Peaks ranked by the sum of the signal density in descending order. Heatmap on CUT&RUN peaks ±3 kb before and after the Region Start Length can be seen below:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6cf52be-1207-4b19-ba02-e6fbfb5441b3",
   "metadata": {},
   "source": [
    "![](images/9.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a323f4e",
   "metadata": {},
   "source": [
    "\n",
    "## 8. Peak-to-Gene Association\n",
    "\n",
    "As a core component of the SWI/SNF remodeling complex, Brg1 exerts its function by binding to chromatin and regulating accessibility to enhancers and promoters of genes, thus facilitating transcriptional activation. Therefore, to infer Brg1’s function from the genomics sites it binds to, we needed to find the genes nearest to these Brg1-bound sites. \n",
    "\n",
    "In order to do so we used the BED files obtained by peak calling to get the gene/region associations and the GREAT algorithm with the following settings:\n",
    "\n",
    "- Assembly: mm10 (GRCm38)\n",
    "- Basal plus extension model\n",
    "- Distal up to 100 kb\n",
    "\n",
    "WT and KO-specific Brg1-associated genes were identified for downstream pathway analysis.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9c90ad4",
   "metadata": {},
   "source": [
    "\n",
    "## 9. WT vs KO Comparison\n",
    "\n",
    "The genes common among WT and KO samples were obtained by entering  the genes we got from GREAT on Venny, a webtool that creates Venn diagrams from lists provided by the user. The overlap analysis showed that the WT samples have 3,796 common peaks (35.8 % overlap). Sample 80.10 WT had 6709 peaks not overlapping with 80.13wheressample 80.13 contained 98 peaks not overlapping with 80.10. The KO samples had 4,722 overlapping peaks (45.5%) with sample 80.7 KO containing 5,171 unique/non-overlapping sites and sample 80.11 KO containing 491 exclusive sites. Τhe later finding strengthens our assumption that despite the lower percentage of DNA mapped to the genome, its DNA contained many peaks. Due to this fact, we didn’t discard this sample from our analysis. The results are also in agreement with the Pearson correlation graph which shows that KO samples are more closely correlated compared to the WT samples. Overall, the overlap between WT replicates and KO replicates is satisfactory for further proceeding with differential enrichment and pathway analysis.\n",
    "\n",
    "After getting the 3796 common genes from our WT samples (80.10 and 80.13) and 4722 KO samples (80.7 and 80.11) we created the final Venn diagram that contains the WT exclusive genes, KO exclusive genes and the common genes between these two:\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4035534-e06c-4521-97c3-5b05400b110f",
   "metadata": {},
   "source": [
    "![](images/10.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "593e75a3",
   "metadata": {},
   "source": [
    "## 10. Integration with publicly available data\n",
    "\n",
    "Genes obtained by GREAT in Enrichr in order to perform enrichment analysis.\n",
    "\n",
    "Genomic regions losing Brg1 occupancy were strongly associated with transcriptional\n",
    "networks governing cell cycle progression and multiciliogenesis, including targets\n",
    "of E2F family transcription factors and the p53/p73 pathway. These findings are\n",
    "consistent with previous evidence that GemC1 cooperates with E2F proteins and p73\n",
    "to drive the multiciliated cell transcriptional program.\n",
    "\n",
    "Regions gaining Brg1 binding were enriched near promoters of genes\n",
    "involved in metabolic regulation, stress-response pathways, and DNA damage\n",
    "checkpoints. The presence of both gained and lost Brg1 binding across shared\n",
    "regulatory pathways (e.g., E2F, MYC, CREM) suggests widespread transcriptional\n",
    "deregulation rather than a simple loss of chromatin engagement.\n",
    "\n",
    "Collectively, these results indicate that GemC1 is required for proper targeting\n",
    "of the SWI/SNF complex to lineage-specific regulatory regions, and that its loss\n",
    "leads to chromatin rewiring incompatible with multiciliated cell differentiation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84a53d5c",
   "metadata": {},
   "source": [
    "## 11. Integration with RNA-seq data\n",
    "\n",
    "A gmt file was created which contained two rows with the exclusive genes on WT and KO samples respectively and loaded it on GseaPreranked option from GSEA. We also used the rnk file that contains the genes extracted from RNA Seq experiment by their log2FoldChange in ascending order.\n",
    "\n",
    "We chose “No_Collapse” on Collapse/Remap to gene symbols field and 10000 on Max size: exclude larger sets field. All other options are set to default.\n",
    "The GSEAPreranked output can be seen below."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa26bce5",
   "metadata": {},
   "source": [
    "![](images/11.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ce01f87-841c-44ee-b341-c34d94ad4be9",
   "metadata": {},
   "source": [
    "We can see that the p-value from both gene sets is high so there is no statistical enrichment."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6557e03c",
   "metadata": {},
   "source": [
    "# 12. Analysis of existing CUT&Tag data\n",
    "\n",
    "For this analysis, we utilized H3K27me3 CUT&Tag data previously generated in the Taraviras lab. CUT&Tag for H3K27me3 was performed in GemC1 WT and KO RGCs, differentiating under the same conditions as the cells from which the CUT&RUN data were derived. \n",
    "\n",
    "Peaks were already called for H3K27me3 data and differential binding was performed using DEseq2. We provided the genes from CUT&Tag experiment in Enrichr using all log2FoldChange from DESeq2 data, as this option gave us the most significant results.\n",
    "\n",
    "Regions losing H3K27me3 were enriched for canonical Polycomb targets and metabolic genes, suggesting disruption of PRC2-dependent repression. In contrast, regions gaining H3K27me3 were associated with genes involved in neuronal differentiation and Hedgehog signaling — pathways critical for early neural specification.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8f3d5b3",
   "metadata": {},
   "source": [
    "For the integration with RNA-seq data again, a gmt file was created which contained 8 rows with 8 different options that we got after filtering the p-value and log2FoldChange columns of the DESeq2 data from our CUT&Tag experiment and loaded it on the GseaPreranked option from GSEA. Again, we used the rnk file that contained the genes extracted from the RNA Seq experiment by their log2FoldChange in ascending order.\n",
    "\n",
    "We chose “No_Collapse” on Collapse/Remap to gene symbols field and 10000 on Max size: exclude larger sets field. All other options are set to default.\n",
    "The gene set that contained all the upregulated genes, without p-value conditions, was the most significant with p-value = 0.005602241:\n",
    "\n",
    "The GSEAPreranked output can be seen below:\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d74cd9e",
   "metadata": {},
   "source": [
    "![](images/12.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e40eda99",
   "metadata": {},
   "source": [
    "The Enrichment plot of the gene set that contains all the K27me3 sites gained in GemC1 KO cells can be seen below. \n",
    "\n",
    "We can see the plot of the Running Enrichment Score (green line) for the gene set as the analysis walks down the ranked gene list. The black lines in the Running Enrichment Score show where the set gene members appear in the ranked list of genes, indicating the leading-edge subset. \n",
    "\n",
    "The Ranked list metric shows the value of log2 fold change as we move down the list of ranked genes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e5ea20f",
   "metadata": {},
   "source": [
    "![](images/13.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "745197b6-f871-4e40-ae5c-2783e6765646",
   "metadata": {},
   "source": [
    "We can see a significant association between gained H3K27me3 and transcriptional downregulation, supporting a functional role for Polycomb retargeting in mediating gene silencing after GemC1 loss."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34d9d0f1-43fe-4ee5-8ab1-0c526f460769",
   "metadata": {},
   "source": [
    "## 13. Conclusions\n",
    "\n",
    "We showed that GemC1 loss reshapes SWI/SNF chromatin targeting and it also regulates Polycomb-mediated repression.\n",
    "\n",
    "Together, these findings support a model in which GemC1 coordinates opposing\n",
    "chromatin regulatory systems:\n",
    "\n",
    "- Promotes SWI/SNF recruitment to activate ciliogenesis programs  \n",
    "- Restricts Polycomb deposition at differentiation-associated loci  \n",
    "\n",
    "Loss of GemC1 disrupts this balance, leading to:\n",
    "\n",
    "- Mislocalization of chromatin remodelers  \n",
    "- Increased deposition of repressive histone marks  \n",
    "- Transcriptional repression of lineage-defining genes  \n",
    "\n",
    "Ultimately, these chromatin alterations likely underlie the impaired\n",
    "differentiation of multiciliated ependymal cells."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
